src/generate_features.py

```diff
@@ -1,3 +1,215 @@
+                 'eval_set', 'order_number', 'order_size', 'user_id']
+    order_types = order_types.join(orders[meta_cols])
+    # strip away metadata and categoricals for clustering
+    X = order_types.drop(columns=['order_dow', 'order_hour_of_day',
+                                  'days_since_prior_order', 'eval_set',
+                                  'order_number', 'user_id']).values
+
+    if config.get('scale_x'):
+        logger.debug('Scaling order features')
+        X = preprocessing.scale(X)
+
+    # filter out rows with NaN vals
+    X = clean(X)
+    logger.debug('Clustering orders based on %d orders and %d variables',
+                 X.shape[0], X.shape[1])
+
+    if config.get('cluster-method') == 'gmm':
+        n_components = config.get('n_components')
+        cov_type = config.get('cov_type')
+        labels = gmm_clust(X, n_components, cov_type)
+    elif config.get('cluster-method') == 'kmeans':
+        k = config.get('k')
+        labels = km_clust(X, k)
+
+    logger.debug('Cluster solution score: %f', score_ch(X, labels))
+    logger.info('Cluster dispersion: %s',
+                np.unique(labels, return_counts=True))
+
+    order_types['label'] = labels
+
+    if config.get('save_cluster_heatmap'):
+        gen_plots(order_types)
+
+    # Limit orders to "prior" set for training data
+    logger.debug('Engineering shopper features')
+    orders = orders[orders['eval_set'] == 'prior']
+    users = pd.DataFrame(orders.groupby('user_id').size().rename('n_orders'))
+
+    # Get num orders per day of week for each user
+    dow_counts = orders.pivot_table(index='user_id', columns='order_dow',
+                                    values='order_number', aggfunc='count')
+    dow_counts.columns = ['n_dow_{}'.format(col)
+                          for col in dow_counts.columns]
+    users = users.join(dow_counts.fillna(0))
+
+    # Get num orders per hour of day for each user
+    hod_counts = orders.pivot_table(index='user_id', columns='order_hour_of_day',
+                                    values='order_number', aggfunc='count')
+    hod_counts.columns = ['n_hod_{}'.format(col) for col in hod_counts.columns]
+    users = users.join(hod_counts.fillna(0))
+
+    # Capture each user's typical order size with summary statistics
+    order_size_stats = orders.groupby('user_id').agg({
+        'order_size': [np.mean, np.std, np.max, np.min]
+        })
+    # Flatten hierarchical column names
+    order_size_stats.columns = ['_'.join(col).strip()
+                                for col in order_size_stats.columns.values]
+
+    users = users.join(order_size_stats)
+    size_count = orders.pivot_table(index='user_id', columns='size_cat',
+                                    values='order_number', aggfunc='count')
+    users = users.join(size_count.fillna(0))
+
+    # Describe users by the composition of their typical order
+    groups = full.groupby('user_id')
```

Comment:
where would you save/display this

Reply:
not sure what you mean here. If you just mean the `groups` object, it is created once and then various operations performed on the groubedby object so that groupby doesn't need to be called multiple times